### ISSUE TITLE
Turn down volume on hotword trigger "Enhancement request "

### ISSUE THREAD

**Issue by ejay-ibm** [eouia/MMM-Hotword#30]
Hi Sean, 

I have a new enhancement request following my volume control request ( that part is done:).
The final objective is to be able to control the volume of the mirror however if a sound is being played (  youtube / raspotify etc ... ) You mentioned in the MMM-AssistantMk2 doc that you are either stopping the voice control while playing a song on youtuve or if the pause option is set to false then trigger works but it stop the video. 

Why not implementing a volume reduce at hotword trigger time ? 
eg: 
hotword Trigger -> get current volume ->lower volume -> wait for the command -> execute command -> put volume as it was.

( That's what the google home assistant do ... i'm not inventing anything)
That would allow reducing the "noise around" to understand the command and continue to play the youtube video.

I hope that I'm clear about what I want to achieve. 

Let me know if you want me to raise another enhancement request on the MMM-AssistantMK2 repository to allow youtube a continuous play while giving a command. 

Have a nice evening. 

Ejay

**Comment by eouia** [comment]
Why I didn't...
First this module is not the alternative Google Home itself. This module is using `Google Assistant SDK for devices` and the purpose of this SDK is giving features of voice(or text) commands and responses for some DIY  or commercial class IoT devices, Not for playing Spotify or Youtube on that device. Some features of Real Google Home are restricted or limited like alarm, news, music...
(Anyway, Google Assistant SDK for devices python library has implemented alarm and news features. That is another story, however)

Second, The real Google Home has Arrayed Mics to catch the voice of user. It can distinguish user's speech from it's own vocal output through it's speaker. But cheap USB mics can't do like that. Sound outputs and your speech could be mixed or confused. That's why I pause one of AMK2 and Hotword to avoid collison of recognition.

Nonetheless, You can do it by yourself. `MMM-NotificationTrigger`( https://github.com/eouia/MMM-NotificationTrigger ) can listen any notification from other modules and translate it into another notifications, or execute shell commands by notification. (shell commands structure is slightly different with that of AMK2)

**Comment by eouia** [comment]
Hmmm... interesting but difficult.
`pauseOnYoutube:false` this option will make your Assistant be live even on youtube playing. But when you order new command during video playing, YT player will be stopped and disappeared. 
For just not quitting YT player even though new command, You can modify line 610 of `MMM-AssistantMk2.js`
```js
  clearResponse() {
    this.subdom.message.innerHTML = ""
    // this.subdom.youtube.innerHTML = ""
    // this.subdom.youtube.style.display = "none"
    // this.youtubePlaying = false
  }
```
But it might make some another side-effect, of course, or not.

I have a trouble with handling RPI ALSA configuration. The main issue is, I cannot figure out how to mix 2 sounds from different process to one speaker output. I'm not that kind of experts so I gave up.
If 2 sounds outputs could be mixed, Your goal could be accomplished.
And for Arrayed Mics, I also cannot understand how it could distinguish. (If I had one, I might have been able to test, but not so.)

Anyway, do you know any good MM module for your purpose (music playing)? Maybe it's better to use that instead my implemented YT Player. My YT Player feature is too deeply combined AMK2 process. I think it is possible to disable my Youtube feature, then to control that MM module with my AMK2 command. Maybe I can add some features to help it.

**Comment by eouia** [comment]
Forget above code. It will not hide YT player after normal ends.

**Comment by ejay-ibm** [comment]
Well thanks you !  it does work as I expected . ( Maybe because I changed in the css file the position the player is very small and goes just bellow the assistant icon ( in the same frame )) 

**
#ASSISTANT_YOUTUBE {
  width:35%;
  height:35%;
  border:none;
  display:none;
  position:center;
  top:0;
  left:0;
  background-color:rgba(0,0,0,0.5);
  z-index:99999;
}
**

![image](https://user-images.githubusercontent.com/38557267/53111605-a4763c00-353d-11e9-8a95-9dc93d6918c0.png)

Anyway, commenting the given code works at least for me. The player continues and the assistant reply during the play. 
![image](https://user-images.githubusercontent.com/38557267/53111632-b6f07580-353d-11e9-8a9f-b977f7c9d293.png)

My next step implement a stop playing feature ( as now their's no way to stop the player ) 
I'll create a new hook for this . I need to figure out then what command need to be sent to stop playing.

And as I said earlier in a last step I'd like to lower the general volume as soon as the hotword is detected and put the volume back as it was once the command is given or better once the command execution end ) 

Thanks for your responsiveness and your help !

**Comment by ejay-ibm** [comment]
regarding the mix of the 2 inputs sound I didn't get any difficulty, if you are struggling with your system maybe I can help . let me know .

**Comment by ejay-ibm** [comment]
Ok I have try to add some code ( and I probably did it wrong ) but I'm trying... 
I'm still missing a part because I don't know how to play with the function.

First the hook  : 
````
 "STOP_PLAY": {
        pattern: "arrête la vidéo",
        command: "STOP_VID"
        }
````

The function : ( yes I copied and paste the code from clear response as it was closing the player at hotword trigger time I thought the same could be used to stop the player on request ) 

````
//added by ejay
  stopPlayer() {
      this.subdom.message.innerHTML = ""
      this.subdom.youtube.innerHTML = ""
      this.subdom.youtube.style.display = "none"
      this.youtubePlaying = false
        }
````

Now between the 2 we need : 

````
  "STOP_VID": {
   <Call the stopPlayer function here>   // I don't know how to achieve that. 
},

````

Thanks for your help .

**Comment by eouia** [comment]
I don't look inside your code deeply.... 
Where did you put your `stopPlayer()` in?
Anyway, you can control module itself with this.

```js
// command section
"STOP_VID" : {
  moduleExec: {
    module: ["MMM-AssistantMk2"],
    exec: (module, params, key) => {
      module.subdom.youtube.innerHTML = ""  // `module` will be `MMM-AssistantMk2` itself by definition of above `module:["MMM-AssistantMk2"]`.
      module.subdom.youtube.style.display = "none"
      module.youtubePlaying = false
    }
  }
}
```
I think this could work, but not tested. If you have a problem, let's solve together.

**Comment by ejay-ibm** [comment]
Thank you seems to be caught now. 

Command is considered but we are getting an error ( It look similar to the one we got with the param earlier ) 

Uncaught TypeError: Cannot read property 'youtube' of undefined
    at Object.exec (MMM-AssistantMk2.js:176)
    at MM.getModules.enumerate (MMM-AssistantMk2.js:767)
    at main.js:433
    at Array.map (<anonymous>)
    at Array.enumerate (main.js:432)
    at AssistantHelper.doCommand (MMM-AssistantMk2.js:764)
    at AssistantHelper.foundHook (MMM-AssistantMk2.js:779)
    at AssistantHelper.conversationEnd (MMM-AssistantMk2.js:796)
    at Class.socketNotificationReceived (MMM-AssistantMk2.js:399)
    at module.js:246

```` line 176 
    module.subdom.youtube.innerHTML = "" 
````

**Comment by eouia** [comment]
I'll test some hours later.

**Comment by eouia** [comment]
change All 3
```js
module.subdom. ...
```
to 
```js
module.assistant.subdom. ...
```

**Comment by ejay-ibm** [comment]
Works like a charm !

````
"STOP_VID" : {
  moduleExec: {
    module: ["MMM-AssistantMk2"],
    exec: (module, params, key) => {
      module.assistant.subdom.youtube.innerHTML = ""  // `module` will be `MMM-AssistantMk2` itself by definition of above `module:["MMM-AssistantMk2"]`.
      module.assistant.subdom.youtube.style.display = "none"
      module.youtubePlaying = false
    }
  },
},
````

Thank you ! 

And now the very last step which was my original request and I'll have a perfect assistant running Turn down the volume once hotword is triggerd and pull it up as it was before starting to answer. 

The workflow would be : 
First Catch current volume and store it : 
````
amixer sget 'PCM' | awk -F"[][]" '{print ""$2""}' | grep %  | awk ' { gsub ( /[%]/, "" ) ;  print }'
````
That will give an integer
 eg : 60 for 6 % 
if we want a string : 
````
amixer sget 'PCM' | awk -F"[][]" '{print ""$2""}' | grep %
````
that will give : 60%

Next step lower the volume at trigger time  : 
````
 amixer sset 'PCM' 5%
````

Wait for the command to be understood  and set the volume back to what it was before answering else if an info is given by the assistant the volume will be too low to understand it. 
````
amixer sset 'PCM'  <Stored Variable from first command > 
````
String would be better to use for the variable as the complete command should look like 
````
amixer sset 'PCM' 60%
````
What do you think about this ? 

I know that is definitly an enhancement request for the assistant But I'll be happy to help you in testing. 
Let me know if this issue should bellong to The Assistant repo ? I can open a new one and put a ref. 

Thanks again 
Ejay

**Comment by ejay-ibm** [comment]
If I'm getting your code right 
I should add 3 additional commands like : 
1
````
"GET_CURRENT_VOL": {
                shellExec: {
                exec: "amixer sget 'PCM' | awk -F"[][]" '{print ""$2""}' | grep %",
    	       // here we need to store the result I don't know how to do that yet
   },
}
````
2
````
"SET_LOW_VOL": {
          shellExec: {
          exec: "amixer sset 'PCM' 5%"
        }
      },
````
3
````
"SET_ORIGINAL_VOL": {
          shellExec: {
          exec: "amixer sset 'PCM'" + Variable string earlier stored here
        }
      },
`````

Now I need to figure out at what place I need to call those commands.

**Comment by eouia** [comment]
Hmmm. I think it's better to make a shell script or MM module to control volume not directly in AMK2. Then control it by AMK2 commands. 
Hmmm... Worthy to try.

**Comment by ejay-ibm** [comment]
You are probably right I'll need to learn a bit more on how notification works  then ... and as you said maybe use the MMM-Notification module then. That will probably take me a couple of weeks to figure out everything but  can try.

**Comment by eouia** [comment]
Give me a few days. I'm starting to build a new module `MMM-Volume` to control volume. It will be easier to use it than to make your own code.

**Comment by ejay-ibm** [comment]
Hey Very cool I'm glad I gave you some Ideas 
I'm searching and trying on my side as well , I'm not a lazy gy waiting for solution and I need to learn  . 
I think i'm almost done with your notification module

**Comment by eouia** [comment]
I finish `MMM-Volume` (https://github.com/eouia/MMM-Volume)
It has all the features you need.(I think)

https://github.com/eouia/MMM-Volume#by-notification
You can control volume with these notifications
`VOLUME_SET`, `VOLUME_STORE`, `VOLUME_RESTORE`.

**Comment by eouia** [comment]
I'll close this issue.
